mllib/src/main/scala/org/apache/spark/mllib/recommendation/MatrixFactorizationModel.scala

```diff
@@ -261,58 +261,93 @@ object MatrixFactorizationModel extends Loader[MatrixFactorizationModel] {
   }
 
   /**
-   * Makes recommendations for all users (or products).
-   * @param rank rank
-   * @param srcFeatures src features to receive recommendations
-   * @param dstFeatures dst features used to make recommendations
-   * @param num number of recommendations for each record
-   * @return an RDD of (srcId: Int, recommendations), where recommendations are stored as an array
-   *         of (dstId, rating) pairs.
+   * Makes recommendations for all users (or items).
+   *
+   * @param rank rank the dimension of the factor vectors
+   * @param srcFactors src factor to receive recommendations
+   * @param dstFactors dst factor used to make recommendations
+   * @param num number of recommendations for each user (or item)
+   * @return an RDD of (srcId, recommendations) pairs, where recommendations are stored as an array
+   *         of (dstId, score) pairs.
    */
-  private def recommendForAll(
+  private[spark] def recommendForAll[T: ClassTag : Fractional](
       rank: Int,
-      srcFeatures: RDD[(Int, Array[Double])],
-      dstFeatures: RDD[(Int, Array[Double])],
-      num: Int): RDD[(Int, Array[(Int, Double)])] = {
-    val srcBlocks = blockify(rank, srcFeatures)
-    val dstBlocks = blockify(rank, dstFeatures)
+      srcFactors: RDD[(Int, Array[T])],
+      dstFactors: RDD[(Int, Array[T])],
+      num: Int): RDD[(Int, Array[(Int, T)])] = {
+    val srcBlocks = blockify(rank, srcFactors)
+    val dstBlocks = blockify(rank, dstFactors)
+    val tag = implicitly[ClassTag[T]].runtimeClass
     val ratings = srcBlocks.cartesian(dstBlocks).flatMap {
       case ((srcIds, srcFactors), (dstIds, dstFactors)) =>
```

Comment:
Agree that this implementation is good for now. You might want to add a TODO comment marking off the portions of the code (blocking and multiplication) that could be replaced with a call to BlockMatrix.